# 2023 Figures

In [135]:
import pandas as pd
import numpy as np

In [12]:
# read in the dataset from Excel, name it fy23
fy23 = pd.read_excel(io="/Users/tainguyen103/Downloads/MASTER Waste 2011-2023.xlsx", 
                     sheet_name="FY23Compost",
                     header=1)

In [13]:
# display the first 5 rows of dataset
fy23.head()

Date Breakfast Lunch Dinner Daily Totals  \
0  2022-07-01 00:00:00         1   NaN      1            2   
1  2022-07-02 00:00:00       NaN     1      2            3   
2        Weekly Totals       NaN   NaN    NaN            0   
3  2022-07-03 00:00:00       NaN   NaN      3            3   
4  2022-07-04 00:00:00         1     2      2            5   

  Lbs per day or week  Lbs/week  
0                 120       NaN  
1                 180       NaN  
2                   0     300.0  
3                 180       NaN  
4                 300       NaN

In [158]:
# display the last 5 rows of dataset. There has been some removals on some last lines on the local excel file to help reflect null values more accurately. 
fy23.tail()

Date Breakfast Lunch      Dinner Daily Totals  \
375  2023-06-29 00:00:00         2     8           6           16   
376  2023-06-30 00:00:00         1     3           3            7   
377        Weekly Totals       NaN   NaN         NaN          nan   
378        GREEN BUCKETS       170   860         846         1876   
379                                       Total Tons                

    Lbs per day or week   Lbs/week  
375                 960        NaN  
376                 420        NaN  
377                 NaN    3240.00  
378              112560  110400.00  
379                          56.28

In [15]:
# count the number of the "Weekly Totals" values in the Date column
week_counts23 = fy23["Date"].value_counts()["Weekly Totals"]
week_counts23

48

### There are 48 weeks during FY23. We are going to multiple this number by 21 since each week has 7 days and there are 3 meals served each day. Notice that there are two special weeks : first one served only for 2 days and last one served for 6 days. So we will subtract 5 + 1 = 6 days (or 18 meals) from the total

In [17]:
# count the total meals of 2023
total_meals23 = 48 * 21 - 6 * 3
print("Total meals in FY23: ",total_meals23)

Total meals in FY23:  990


In [159]:
# count the number of null values for breakfast, subtract it by 48 since we do not count the null values of "Weekly Totals"
null_bf23 = fy23["Breakfast"].isnull().value_counts()[True] - 48 

In [20]:
# count the number of null values for lunch, subtract it by 48 since we do not count the null values of "Weekly Totals"
null_ln23 = fy23["Lunch"].isnull().value_counts()[True] - 48

In [21]:
# count the number null values for dinner, subtract it by 48 since we do not count the null values of "Weekly Totals"
null_dn23 = fy23["Dinner"].isnull().value_counts()[True] - 48

In [22]:
# print out total number of null values for all 3 meals
print(null_bf23 + null_ln23 + null_dn23)

434


In [50]:
# calculate the percentage of null meals in total meals
perc23 = (434/990)*100
perc23

43.83838383838384

In [53]:
# count the number of green buckets
total_gb23 = fy23["Daily Totals"].iloc[:378].sum()
total_gb23

1876

In [54]:
# count the number of days on which meals are served
total_days23 = 48 * 7 - 6 
total_days23

330

In [96]:
# calculate the average number of green buckets for each day
average_gb23 = total_gb23 / total_days23
average_gb23

5.684848484848485

In [170]:
# filter by dates that used below the average number of green bucket, which is less than or equal to 5.68
fy23['Daily Totals'] = fy23['Daily Totals'].astype(str)
daycount23_belowAVG = fy23.loc[fy23['Daily Totals'].isin(['1','2', '3', '4', '5'])]
daycount23_belowAVG

Date Breakfast Lunch Dinner Daily Totals  \
0    2022-07-01 00:00:00         1   NaN      1            2   
1    2022-07-02 00:00:00       NaN     1      2            3   
3    2022-07-03 00:00:00       NaN   NaN      3            3   
4    2022-07-04 00:00:00         1     2      2            5   
5    2022-07-05 00:00:00         1   NaN      1            2   
..                   ...       ...   ...    ...          ...   
361  2023-06-17 00:00:00         1     3    NaN            4   
363  2023-06-18 00:00:00       NaN     2    NaN            2   
366  2023-06-21 00:00:00         1     2      2            5   
367  2023-06-22 00:00:00         1     2      2            5   
369  2023-06-24 00:00:00         1     2      1            4   

    Lbs per day or week  Lbs/week  
0                   120       NaN  
1                   180       NaN  
3                   180       NaN  
4                   300       NaN  
5                   120       NaN  
..                  ...       ...  
361                 240       NaN  
363                 120       NaN  
366                 300       NaN  
367                 300       NaN  
369                 240       NaN  

[117 rows x 7 columns]

### According to the filtered table we can see that there are 117 rows (or dates) with the number of green buckets used falling below the average. 

In [154]:
# filter by dates where all 3 meals are null or do not have a recorded value
nullallmeals23 = fy23.loc[(fy23['Breakfast'].isnull()) & (fy23['Lunch'].isnull()) & (fy23['Dinner'].isnull())]
actual23 = nullallmeals23.loc[nullallmeals23["Date"] != "Weekly Totals"]
actual23

Date Breakfast Lunch Dinner Daily Totals  \
11   2022-07-10 00:00:00       NaN   NaN    NaN            0   
17   2022-07-16 00:00:00       NaN   NaN    NaN            0   
24   2022-07-22 00:00:00       NaN   NaN    NaN            0   
27   2022-07-24 00:00:00       NaN   NaN    NaN            0   
28   2022-07-25 00:00:00       NaN   NaN    NaN            0   
29   2022-07-26 00:00:00       NaN   NaN    NaN            0   
31   2022-07-28 00:00:00       NaN   NaN    NaN            0   
32   2022-07-29 00:00:00       NaN   NaN    NaN            0   
33   2022-07-30 00:00:00       NaN   NaN    NaN            0   
35   2022-07-31 00:00:00       NaN   NaN    NaN            0   
41   2022-08-06 00:00:00       NaN   NaN    NaN            0   
43   2022-08-07 00:00:00       NaN   NaN    NaN            0   
44   2022-08-08 00:00:00       NaN   NaN    NaN            0   
45   2022-08-09 00:00:00       NaN   NaN    NaN            0   
59   2022-08-21 00:00:00       NaN   NaN    NaN            0   
67   2022-08-28 00:00:00       NaN   NaN    NaN            0   
75   2022-09-04 00:00:00       NaN   NaN    NaN            0   
81   2022-09-10 00:00:00       NaN   NaN    NaN            0   
83   2022-09-11 00:00:00       NaN   NaN    NaN            0   
89   2022-09-17 00:00:00       NaN   NaN    NaN            0   
91   2022-09-18 00:00:00       NaN   NaN    NaN            0   
97   2022-09-24 00:00:00       NaN   NaN    NaN            0   
99   2022-09-25 00:00:00       NaN   NaN    NaN            0   
115  2022-10-09 00:00:00       NaN   NaN    NaN            0   
129  2022-10-22 00:00:00       NaN   NaN    NaN            0   
131  2022-10-23 00:00:00       NaN   NaN    NaN            0   
137  2022-10-29 00:00:00       NaN   NaN    NaN            0   
147  2022-11-06 00:00:00       NaN   NaN    NaN            0   
166  2022-11-23 00:00:00       NaN   NaN    NaN            0   
167  2022-11-24 00:00:00       NaN   NaN    NaN            0   
168  2022-11-25 00:00:00       NaN   NaN    NaN            0   
169  2022-11-26 00:00:00       NaN   NaN    NaN            0   
171  2022-11-27 00:00:00       NaN   NaN    NaN            0   
177  2022-12-03 00:00:00       NaN   NaN    NaN            0   
179  2022-12-04 00:00:00       NaN   NaN    NaN            0   
185  2022-12-10 00:00:00       NaN   NaN    NaN            0   
192  2022-12-16 00:00:00       NaN   NaN    NaN            0   
193  2022-12-17 00:00:00       NaN   NaN    NaN            0   
195  2023-01-15 00:00:00       NaN   NaN    NaN            0   
196  2023-01-16 00:00:00       NaN   NaN    NaN            0   
197  2023-01-17 00:00:00       NaN   NaN    NaN            0   
200  2023-01-20 00:00:00       NaN   NaN    NaN            0   
201  2023-01-21 00:00:00       NaN   NaN    NaN            0   
203  2023-01-22 00:00:00       NaN   NaN    NaN            0   
227  2023-02-12 00:00:00       NaN   NaN    NaN            0   
257  2023-03-11 00:00:00       NaN   NaN    NaN            0   
259  2023-03-19 00:00:00       NaN   NaN    NaN            0   
275  2023-04-02 00:00:00       NaN   NaN    NaN            0   
307  2023-04-30 00:00:00       NaN   NaN    NaN            0   
332  2023-05-22 00:00:00       NaN   NaN    NaN            0   
336  2023-05-26 00:00:00       NaN   NaN    NaN            0   
337  2023-05-27 00:00:00       NaN   NaN    NaN            0   
339  2023-05-28 00:00:00       NaN   NaN    NaN            0   
340  2023-05-29 00:00:00       NaN   NaN    NaN            0   
345  2023-06-03 00:00:00       NaN   NaN    NaN            0   
347  2023-06-04 00:00:00       NaN   NaN    NaN            0   
351  2023-06-08 00:00:00       NaN   NaN    NaN            0   

    Lbs per day or week  Lbs/week  
11                    0       NaN  
17                    0       NaN  
24                    0       NaN  
27                    0       NaN  
28                    0       NaN  
29                    0       NaN  
31                    0       NaN  
32                 

In [161]:
# count the days, or number of rows
actual23.shape[0]

57

# 2022 Figures 

In [45]:
# read in the dataset from Excel, name it fy22
fy22 = pd.read_excel(io="/Users/tainguyen103/Downloads/MASTER Waste 2011-2023.xlsx", 
                     sheet_name="FY22Compost",
                     header=1)

In [26]:
# display the first 5 rows of dataset
fy22.head()

Date Breakfast Lunch Dinner Daily Totals  \
0  2021-07-04 00:00:00       NaN     2      2            4   
1  2021-07-05 00:00:00       NaN   NaN    NaN            0   
2  2021-07-06 00:00:00       NaN     1      3            4   
3  2021-07-07 00:00:00       NaN   NaN      2            2   
4  2021-07-08 00:00:00       NaN     2      3            5   

  Lbs per day or week Lbs/week  Unnamed: 7  Unnamed: 8               Date.1  \
0                 240                  NaN         NaN  2021-07-04 00:00:00   
1                   0                  NaN         NaN  2021-07-05 00:00:00   
2                 240                  NaN         NaN  2021-07-06 00:00:00   
3                 120                  NaN         NaN  2021-07-07 00:00:00   
4                 300                  NaN         NaN  2021-07-08 00:00:00   

  Breakfast.1 Lunch.1 Dinner.1 Daily Totals.1  Lbs per day or week.1  \
0                                           0                    0.0   
1                                           0                    0.0   
2                                           0                    0.0   
3                                           0                    0.0   
4                                           0                    0.0   

   Weekly Totals   
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN

In [162]:
# display the last 5 rows of dataset
fy22.tail()

Date Breakfast Lunch      Dinner Daily Totals  \
405      Weekly Totals                                            
406      GREEN BUCKETS       326   789         822         1937   
407     YELLOW BUCKETS                                            
408  TOTAL ALL BUCKETS                  Total lbs.          nan   
409                                     Total Tons                

    Lbs per day or week  Lbs/week  Unnamed: 7  Unnamed: 8 Date.1 Breakfast.1  \
405                 NaN       900         NaN         NaN    NaN         NaN   
406              116220    116220         NaN         NaN    NaN         NaN   
407              1039.5    1039.5         NaN         NaN    NaN         NaN   
408            117259.5  117259.5         NaN         NaN    NaN         NaN   
409                      58.62975         NaN         NaN    NaN         NaN   

    Lunch.1 Dinner.1 Daily Totals.1  Lbs per day or week.1  Weekly Totals   
405     NaN      NaN            NaN                    NaN             NaN  
406     NaN      NaN            NaN                    NaN             NaN  
407     NaN      NaN            NaN                    NaN             NaN  
408     NaN      NaN            NaN                    NaN             NaN  
409     NaN      NaN            NaN                    NaN             NaN

In [163]:
# count the number of the "Weekly Totals" values in the Date column
week_counts22 = fy22["Date"].value_counts()["Weekly Totals"]
week_counts22

51

### There are 51 weeks during FY22. We are going to multiple this number by 21 since each week has 7 days and there are 3 meals served each day. Notice that the last week only served for 2 days. So we will subtract 2 days (or 6 meals) from the total

In [164]:
# count the total meals in 2022
total_meals22 = 51 * 21 - 2 * 3
print("Total meals in FY22: ", total_meals22)

Total meals in FY22:  1065


In [47]:
# count the number of null values for breakfast, subtract it by 51 since we do not count the null values of "Weekly Totals"
null_bf22 = fy22["Breakfast"].isnull().value_counts()[True] - 51

In [38]:
# count the number of null values for breakfast, subtract it by 51 since we do not count the null values of "Weekly Totals"
null_ln22 = fy22["Lunch"].isnull().value_counts()[True] - 51

In [39]:
# count the number of null values for breakfast, subtract it by 51 since we do not count the null values of "Weekly Totals"
null_dn22 = fy22["Dinner"].isnull().value_counts()[True] - 51

In [165]:
# print out total number of null values for all 3 meals
print(null_bf22 + null_ln22 + null_dn22)

400


In [166]:
# calculate the percentage of null meals in total meals
perc22 = (400/1065) * 100
perc22

37.558685446009385

In [93]:
# count the number of days on which meals are served
total_days22 = 51 * 7 - 2
total_days22

355

In [167]:
# calculate the average number of green buckets for each day
average_gb22 = 1937 / total_days22 #1937 is the recorded total from the excel sheet
average_gb22

5.456338028169014

In [169]:
# filter by dates that used below the average number of green bucket, which is less than or equal to 5.45
fy22['Daily Totals'] = fy22['Daily Totals'].astype(str)
daycount22_belowAVG = fy22.loc[fy22['Daily Totals'].isin(['1','2', '3', '4', '5'])]
daycount22_belowAVG

Date Breakfast Lunch Dinner Daily Totals  \
0    2021-07-04 00:00:00       NaN     2      2            4   
2    2021-07-06 00:00:00       NaN     1      3            4   
3    2021-07-07 00:00:00       NaN   NaN      2            2   
4    2021-07-08 00:00:00       NaN     2      3            5   
5    2021-07-09 00:00:00         2     1      2            5   
..                   ...       ...   ...    ...          ...   
398  2022-06-25 00:00:00       NaN     1    NaN            1   
400  2022-06-26 00:00:00       NaN     2      2            4   
402  2022-06-28 00:00:00         2     3    NaN            5   
403  2022-06-29 00:00:00         1     2    NaN            3   
404  2022-06-30 00:00:00         1     2    NaN            3   

    Lbs per day or week Lbs/week  Unnamed: 7  Unnamed: 8               Date.1  \
0                   240                  NaN         NaN  2021-07-04 00:00:00   
2                   240                  NaN         NaN  2021-07-06 00:00:00   
3                   120                  NaN         NaN  2021-07-07 00:00:00   
4                   300                  NaN         NaN  2021-07-08 00:00:00   
5                   300      NaN         NaN         NaN  2021-07-09 00:00:00   
..                  ...      ...         ...         ...                  ...   
398                  60                  NaN         NaN                  NaN   
400                 240                  NaN         NaN                  NaN   
402                 300                  NaN         NaN                  NaN   
403                 180                  NaN         NaN                  NaN   
404                 180                  NaN         NaN                  NaN   

    Breakfast.1 Lunch.1 Dinner.1 Daily Totals.1  Lbs per day or week.1  \
0                                             0                    0.0   
2                                             0                    0.0   
3                                             0                    0.0   
4                                             0                    0.0   
5                                             0                    0.0   
..          ...     ...      ...            ...                    ...   
398         NaN     NaN      NaN            NaN                    NaN   
400         NaN     NaN      NaN            NaN                    NaN   
402         NaN     NaN      NaN            NaN                    NaN   
403         NaN     NaN      NaN            NaN                    NaN   
404         NaN     NaN      NaN            NaN                    NaN   

     Weekly Totals   
0               NaN  
2               NaN  
3               NaN  
4               NaN  
5               NaN  
..              ...  
398             NaN  
400             NaN  
402             NaN  
403             NaN  
404             NaN  

[113 rows x 16 columns]

### According to the filtered table we can see that there are 113 rows (or dates) with the number of green buckets used falling below the average.

In [156]:
# filter by dates where all 3 meals are null or do not have a recorded value
nullallmeals22 = fy22.loc[(fy22['Breakfast'].isnull()) & (fy22['Lunch'].isnull()) & (fy22['Dinner'].isnull())]
actual22 = nullallmeals22.loc[nullallmeals22["Date"] != "Weekly Totals"]
actual22

Date Breakfast Lunch Dinner Daily Totals  \
1    2021-07-05 00:00:00       NaN   NaN    NaN            0   
30   2021-07-31 00:00:00       NaN   NaN    NaN            0   
32   2021-08-01 00:00:00       NaN   NaN    NaN            0   
40   2021-08-08 00:00:00       NaN   NaN    NaN            0   
72   2021-09-05 00:00:00       NaN   NaN    NaN            0   
..                   ...       ...   ...    ...          ...   
378  2022-06-07 00:00:00       NaN   NaN    NaN            0   
382  2022-06-11 00:00:00       NaN   NaN    NaN            0   
384  2022-06-12 00:00:00       NaN   NaN    NaN            0   
390  2022-06-18 00:00:00       NaN   NaN    NaN            0   
401  2022-06-27 00:00:00       NaN   NaN    NaN            0   

    Lbs per day or week Lbs/week  Unnamed: 7  Unnamed: 8               Date.1  \
1                     0                  NaN         NaN  2021-07-05 00:00:00   
30                    0                  NaN         NaN  2021-07-31 00:00:00   
32                    0                  NaN         NaN  2021-08-01 00:00:00   
40                    0                  NaN         NaN  2021-08-08 00:00:00   
72                    0                  NaN         NaN  2021-09-05 00:00:00   
..                  ...      ...         ...         ...                  ...   
378                   0                  NaN         NaN                  NaN   
382                   0                  NaN         NaN                  NaN   
384                   0                  NaN         NaN                  NaN   
390                   0                  NaN         NaN                  NaN   
401                   0                  NaN         NaN                  NaN   

    Breakfast.1 Lunch.1 Dinner.1 Daily Totals.1  Lbs per day or week.1  \
1                                             0                    0.0   
30                                            0                    0.0   
32                                            0                    0.0   
40                                            0                    0.0   
72          NaN                               0                    0.0   
..          ...     ...      ...            ...                    ...   
378         NaN     NaN      NaN            NaN                    NaN   
382         NaN     NaN      NaN            NaN                    NaN   
384         NaN     NaN      NaN            NaN                    NaN   
390         NaN     NaN      NaN            NaN                    NaN   
401         NaN     NaN      NaN            NaN                    NaN   

     Weekly Totals   
1               NaN  
30              NaN  
32              NaN  
40              NaN  
72              NaN  
..              ...  
378             NaN  
382             NaN  
384             NaN  
390             NaN  
401             NaN  

[108 rows x 16 columns]

In [171]:
# count the days, or number of rows
actual22.shape[0]

108